### DISCUSSION TITLE
How do I use a custom ESLint plugin in the CliEngine class

### DISCUSSION THREAD

**Discussion by xinqiymsz** [eslint/eslint#14196] [Q&A] [answered]
I want to use the CliEngine class to see an array of ESLint error messages. The CliEngine class wants to import my custom rules. Which argument in the CliEngine class should I use to import my custom ESLint plugin

```
const cli = new CLIEngine({???}) `
const reports = cli.executeOnFiles(paths).results
```

**Comment by mdjermanovic** [comment, marked as answer]
Hi @xinqiymsz!

Assuming you have `eslint-plugin-myplugin` installed in `node_modules`, and that the plugin provides `myplugin/my-rule` rule, there are several ways to load the plugin and enable the rule:

* `new CLIEngine({ plugins: ["myplugin"], rules: { "myplugin/my-rule": ["error"] } })`
* `new CLIEngine({ baseConfig: { plugins: ["myplugin"], rules: { "myplugin/my-rule": ["error"] } } })`
* Or, make an `.eslintrc.*` config file with `{ plugins: ["myplugin"], rules: { "myplugin/my-rule": ["error"] } }` configuration and then just use `new CLIEngine()`. ESLint with read configuration from the file.

**Reply by xinqiymsz** [reply]
My custom ESLint plugin is eslint-plugin-code-complexity, do you mean to write like this?
```
const cli = new CLIEngine({
    plugins: ["code-complexity"],
    rules: {
        "code-complexity/code-complexity": ["error", { max: 1 }]
    },
})
```

**Reply by xinqiymsz** [reply]
But it didn't work

**Reply by xinqiymsz** [reply]
> Hi @xinqiymsz!
> 
> Assuming you have `eslint-plugin-myplugin` installed in `node_modules`, and that the plugin provides `myplugin/my-rule` rule, there are several ways to load the plugin and enable the rule:
> 
> * `new CLIEngine({ plugins: ["myplugin"], rules: { "myplugin/my-rule": ["error"] } })`
> * `new CLIEngine({ baseConfig: { plugins: ["myplugin"], rules: { "myplugin/my-rule": ["error"] } } })`
> * Or, make an `.eslintrc.*` config file with `{ plugins: ["myplugin"], rules: { "myplugin/my-rule": ["error"] } }` configuration and then just use `new CLIEngine()`. ESLint with read configuration from the file.

Could you please help me to check why it is not effective? Thank you!!
My custom ESLint plugin is eslint-plugin-code-complexity, do you mean to write like this?

const cli = new CLIEngine({
    plugins: ["code-complexity"],
    rules: {
        "code-complexity/code-complexity": ["error", { max: 1 }]
    },
})

**Reply by mdjermanovic** [reply]
I think the plugin exports a rule named `codecomplexity` (without `-`), so it would be:

```js
rules: {
    "code-complexity/codecomplexity": ["error", { max: 1 }]
}
```

**Reply by xinqiymsz** [reply]
> I think the plugin exports a rule named `codecomplexity` (without `-`), so it would be:
> 
> ```js
> rules: {
>     "code-complexity/codecomplexity": ["error", { max: 1 }]
> }
> ```

I changed the name of my rule to codecomplexity and my plugin ID is code-complexity, so I wrote it this way but it still doesn't work
```
// package.json
"dependencies": {
    "eslint-plugin-code-complexity": "0.0.3",
  },
// This is the project file where I introduced the plugin
const eslint = require("eslint")
const glob = require("glob")
const com = require("eslint-plugin-code-complexity")
const { CLIEngine } = eslint

const DEFAULT_IGNORE_PATTERNS = ["node_modules/**", "build/**", "dist/**", "output/**", "common_build/**"]

const cli = new CLIEngine({
    plugins: ["code-complexity"],
    rules: { "code-complexity/codecomplexity": ["error", { max: 1 }] } 
}) //The cli variable is always empty

```
>The cli variable is always empty


This is my ESLint plugin file directory
![截屏2021-03-16 上午11 10 54](https://user-images.githubusercontent.com/32388393/111250454-506cb180-8648-11eb-9a1d-e28870c8e88b.png)

The cli variable is always empty, could you please download my eslint-plugin-code-complexity plugin on NPM to take a look at why the CLI is empty? This problem has been troubling me for several days. Thank you very much 🙏

**Reply by mdjermanovic** [reply]
By "empty" you mean there are no linting results?

`new CLIEngine()` just initializes the `CLIEngine` and sets configuration. It doesn't lint anything yet. You'll have to run it on some files or text to lint: [CLIEngine#executeOnFiles()](https://eslint.org/docs/developer-guide/nodejs-api#cliengineexecuteonfiles) or [CLIEngine#executeOnText()](https://eslint.org/docs/developer-guide/nodejs-api#cliengineexecuteontext).

**Reply by xinqiymsz** [reply]
> By "empty" you mean there are no linting results?
> 
> `new CLIEngine()` just initializes the `CLIEngine` and sets configuration. It doesn't lint anything yet. You'll have to run it on some files or text to lint: [CLIEngine#executeOnFiles()](https://eslint.org/docs/developer-guide/nodejs-api#cliengineexecuteonfiles) or [CLIEngine#executeOnText()](https://eslint.org/docs/developer-guide/nodejs-api#cliengineexecuteontext).


No, I typed the command in the root directory of the project, and the file runs. Here is my code. The CLI variable is always undefined. I think it is a syntax problem, you can try it out, or you have a reference to the custom plugin code use case in new CliEngine (), can you send me a message? thanks!
![截屏2021-03-18 上午11 08 11](https://user-images.githubusercontent.com/32388393/111567767-75912980-87da-11eb-805d-176f043e14d6.png)

```
const eslint = require("eslint")
const glob = require("glob")
const { CLIEngine } = eslint

const DEFAULT_IGNORE_PATTERNS = ["node_modules/**", "build/**", "dist/**", "output/**", "common_build/**"]

const cli = new CLIEngine({
    plugins: ["code-complexity"],
    rules: { "code-complexity/codecomplexity": ["error", { max: 10 }] },
})
console.log(cli, '//////');
const REG_FUNC_TYPE = /^(Method |Async function |Arrow function |Function )/g
const MESSAGE_PREFIX = "Maximum allowed is 1."
const MESSAGE_SUFFIX = "has a complexity of "
function getFunctionType(message) {
    let a = message.match(REG_FUNC_TYPE) || []
    return a.length ? a[0] : "--"
}

function getFileName(filePath) {
    return filePath.replace(process.cwd(), "").trim()
}

function executeOnFiles(paths, min) {
    const reports = cli.executeOnFiles(paths).results
    const result = []
    const fileCount = paths.length
    let funcCount = 0
    for (let i = 0; i < reports.length; i++) {
        const { messages, filePath } = reports[i]
        for (let j = 0; j < messages.length; j++)
          
            }
        }
    }
    return { normalRatio, slightRatio, seriousRatio, fileCount, funcCount, result }
}

function getFiles() {
    return new Promise(resolve => {
        glob(
            "**/*.?(vue|js|jsx)",
            {
                ignore: DEFAULT_IGNORE_PATTERNS
            },
            function(er, files) {
                resolve(files)
              
            }
        )
    })
}
module.exports = async function(param) {
    // options is optional
    const { min = 1 } = param || {}

    let files = await getFiles()

    return executeOnFiles(files, min)
}
```

**Reply by xinqiymsz** [reply]
> By "empty" you mean there are no linting results?
> 
> `new CLIEngine()` just initializes the `CLIEngine` and sets configuration. It doesn't lint anything yet. You'll have to run it on some files or text to lint: [CLIEngine#executeOnFiles()](https://eslint.org/docs/developer-guide/nodejs-api#cliengineexecuteonfiles) or [CLIEngine#executeOnText()](https://eslint.org/docs/developer-guide/nodejs-api#cliengineexecuteontext).

Can you give an example of a custom plugin referenced in the CliEngine class that works successfully? thank you !!🙏

**Reply by mdjermanovic** [reply]
It isn't undefined. It's an object without own properties, methods are on the prototype.

Try to run `cli.executeOnFiles()` with some paths, and it will return linting result.

**Reply by xinqiymsz** [reply]
> It isn't undefined. It's an object without own properties, methods are on the prototype.
> 
> Try to run `cli.executeOnFiles()` with some paths, and it will return linting result.

I tried passing in the path, but it still doesn't work. I don't know why?
The CliEngine class works with its own ESLint rules, but custom rules do not work.
I hope you can try it. Thank you🙏

**Reply by mdjermanovic** [reply]
`test.js`:

```js
() => {
    1 || 2 || 3 || 4 || 5 || 6 || 7 || 8 || 9 || 10 || 11;
}

```

`check.js`:

```js
const { CLIEngine } = require("eslint");

const cli = new CLIEngine({
    plugins: ["code-complexity"],
    rules: { "code-complexity/codecomplexity": ["error", { max: 10 }] },
});

const lintResults = cli.executeOnFiles("test.js").results;

const lintResult = lintResults[0];

const lintMessages = lintResult.messages;

const lintMessage = lintMessages[0];

console.log(lintMessage);
```

run:

```
node check.js
```

prints:

```
{
  ruleId: 'code-complexity/codecomplexity',
  severity: 2,
  message: 'Arrow function has a complexity of 11. Maximum allowed is 10.',
  line: 1,
  column: 1,
  nodeType: 'ArrowFunctionExpression',
  messageId: 'complex',
  endLine: 3,
  endColumn: 2
}
```

**Reply by xinqiymsz** [reply]
> ```js
> () => {
>     1 || 2 || 3 || 4 || 5 || 6 || 7 || 8 || 9 || 10 || 11;
> }
> ```

Maybe it's my path reference problem. I'll check it. Thank you very much indeed🙏🙏
